src/csg_tree.cpp

```diff
@@ -566,76 +444,188 @@ void CsgOpNode::BatchUnion() const {
-              std::dynamic_pointer_cast<CsgLeafNode>(children_[start + j]));
+              std::dynamic_pointer_cast<CsgLeafNode>(children[start + j]));
         }
         impls.push_back(
             std::make_shared<const Manifold::Impl>(CsgLeafNode::Compose(tmp)));
       }
     }
 
-    children_.erase(children_.begin() + start, children_.end());
-    children_.push_back(
+    children.erase(children.begin() + start, children.end());
+    children.push_back(
         std::make_shared<CsgLeafNode>(BatchBoolean(OpType::Add, impls)));
     // move it to the front as we process from the back, and the newly added
     // child should be quite complicated
-    std::swap(children_.front(), children_.back());
+    std::swap(children.front(), children.back());
   }
 }
 
-/**
- * Flatten the children to a list of leaf nodes and return them.
- * If forceToLeafNodes is true, the list will be guaranteed to be a list of leaf
- * nodes (i.e. no ops). Otherwise, the list may contain ops. Note that this
- * function will not apply the transform to children, as they may be shared with
- * other nodes.
- */
-std::vector<std::shared_ptr<CsgNode>> &CsgOpNode::GetChildren(
-    bool forceToLeafNodes) const {
+CsgOpNode::CsgOpNode() {}
+
+CsgOpNode::CsgOpNode(const std::vector<std::shared_ptr<CsgNode>> &children,
+                     OpType op)
+    : impl_(Impl{}), op_(op) {
   auto impl = impl_.GetGuard();
+  impl->children_ = children;
+}
 
-  if (forceToLeafNodes && !impl->forcedToLeafNodes_) {
-    impl->forcedToLeafNodes_ = true;
-    for_each(ExecutionPolicy::Par, impl->children_.begin(),
-             impl->children_.end(), [](auto &child) {
-               if (child->GetNodeType() != CsgNodeType::Leaf) {
-                 child = child->ToLeafNode();
-               }
-             });
-  }
-  return impl->children_;
+CsgOpNode::CsgOpNode(std::vector<std::shared_ptr<CsgNode>> &&children,
+                     OpType op)
+    : impl_(Impl{}), op_(op) {
+  auto impl = impl_.GetGuard();
+  impl->children_ = children;
 }
 
-void CsgOpNode::SetOp(OpType op) {
-  switch (op) {
-    case OpType::Add:
-      op_ = CsgNodeType::Union;
-      break;
-    case OpType::Subtract:
-      op_ = CsgNodeType::Difference;
-      break;
-    case OpType::Intersect:
-      op_ = CsgNodeType::Intersection;
-      break;
+std::shared_ptr<CsgNode> CsgOpNode::Boolean(
+    const std::shared_ptr<CsgNode> &second, OpType op) {
+  std::vector<std::shared_ptr<CsgNode>> children;
+  children.push_back(shared_from_this());
+  children.push_back(second);
+
+  return std::make_shared<CsgOpNode>(children, op);
+}
+
+std::shared_ptr<CsgNode> CsgOpNode::Transform(const mat3x4 &m) const {
+  auto node = std::make_shared<CsgOpNode>();
+  node->impl_ = impl_;
+  node->transform_ = m * Mat4(transform_);
+  node->op_ = op_;
+  return node;
+}
+
+struct CsgStackFrame {
+  bool finalize;
+  OpType parent_op;
+  mat3x4 transform;
+  std::vector<std::shared_ptr<CsgLeafNode>> *parent_children;
```

Comment:
what's the relationship between `parent_children` and `left_children` and `right_children`? And why is just the parent a pointer?

Reply:
`parent_children` is the list of children of the parent (or ancestor), while `left_children` and `right_children` are the sets of children of the current node.

Will explain it better with some comment.

Reply:
I think I might call `parent_children` `siblings`, if I'm understanding you correctly?